src/machine-c-api.h

```diff
@@ -241,6 +241,54 @@ typedef struct {          // NOLINT(modernize-use-using)
     bool yield_automatic; ///< Make yield automatic available?
 } cm_htif_config;
 
+/// \brief VirtIO device type
+typedef enum { // NOLINT(modernize-use-using)
+    CM_VIRTIO_DEVICE_NONE = 0,
+    CM_VIRTIO_DEVICE_CONSOLE,
+    CM_VIRTIO_DEVICE_P9FS,
+    CM_VIRTIO_DEVICE_NET_USER,
+    CM_VIRTIO_DEVICE_NET_TUNTAP
+} CM_VIRTIO_DEVICE_TYPE;
+
+/// \brief VirtIO console device state configuration
+typedef struct {
+} cm_virtio_console_config;
+
+/// \brief VirtIO Plan 9 filesystem device state configuration
+typedef struct {
+    const char *mount_tag;   ///< Guest mount tag
+    const char *shared_path; ///< Path to the host shared directory
+} cm_virtio_p9fs_config;
+
+/// \brief VirtIO user network device state configuration
+typedef struct {
+} cm_virtio_net_user_config;
+
+/// \brief VirtIO TUN/TAP network device state configuration
+typedef struct {
+    const char *iface; ///< Host's tap network interface (e.g "tap0")
+} cm_virtio_net_tuntap_config;
+
+/// \brief VirtIO device union
+typedef union {
+    cm_virtio_console_config console;       ///< Console
+    cm_virtio_p9fs_config p9fs;             ///< Plan 9 filesystem
+    cm_virtio_net_user_config net_user;     ///< User-mode networking
+    cm_virtio_net_tuntap_config net_tuntap; ///< TUN/TAP networking
+} cm_virtio_device_config_union;
+
+/// \brief VirtIO device state configuration
+typedef struct {
+    CM_VIRTIO_DEVICE_TYPE type;           ///< VirtIO device type
+    cm_virtio_device_config_union device; ///< VirtIO device config
```

Comment:
What is your opinion about using a nameless union here to avoid writing "device" all the time?

Reply:
I am against using nameless union or structs, because there are not friendly with C binding generators. Also it's not part of C99 spec, they were introduced in C11. And I want the C API to be compatible with C99.